_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,894 @@
+
+We create a `Request` class and use that to build our mail. Within the response variable we call `setMethod()`, `setEndpoint()`, and `setBody()` methods on the `request` class and return `sg.api(request)` as a `Try[Response]` which will hold the response from the SendGrid server.
+
+Finally, we pattern match on `response`, in case our mail was sent successfully, we'll print the response body, status code, and headers in our console for debugging purposes and reply with a message telling the user to check their mail to complete the sign in process, otherwise we print out the error message and tell the user to try again later.
+
+Here is the full code.
+
+```scala
+package com.rockthejvm
+
+import com.sendgrid.{SendGrid, Method, Request, Response}
+import com.sendgrid.helpers.mail.objects.{Email, Content, Attachments}
+import com.sendgrid.helpers.mail.Mail
+import java.nio.file.{Paths, Files}
+import java.util.Base64
+import scala.util.*
+import org.http4s.dsl.io.*
+import cats.effect.IO
+
+class EmailService(sendto: String):
+    val from = new Email("hkateu@gmail.com")
+    val to = new Email(sendto)
+
+    val subject = "Sending with SendGrid is Fun"
+    val message = s"""
+        |<H3>Website Login</H3>
+        |<p>Please complete the login process using Google Authenticator</p>
+        |<p>Scan the QR Barcode attached and send us the token number.</p>
+        """.stripMargin
+
+    val file = Paths.get("barCode.png")
+    val content = new Content("text/html", message)
+
+    val mail = new Mail(from, subject, to, content)
+    val attachments = new Attachments()
+    attachments.setFilename(file.getFileName().toString())
+    attachments.setType("application/png")
+    attachments.setDisposition("attachment")
+    val attachmentContentBytes = Files.readAllBytes(file)
+    val attachmentContent = Base64.getEncoder().encodeToString(attachmentContentBytes)
+    attachments.setContent(attachmentContent)
+    mail.addAttachments(attachments)
+
+    val sg = new SendGrid(Properties.envOrElse("SENDGRID_API_KEY","undefined"))
+    val request = new Request()
+    val response: Try[Response] = Try {
+        request.setMethod(Method.POST)
+        request.setEndpoint("mail/send")
+        request.setBody(mail.build())
+        val res = sg.api(request)
+        res
+    }
+
+    val checkStatus = response match
+        case Success(res) =>
+            IO(println(res.getBody())) >>
+            IO(println(res.getStatusCode())) >>
+            IO(println(res.getHeaders())) >>
+            Ok("We sent you an email, please follow the steps to complete the signin process.")
+        case Failure(ex) =>
+            IO(println(ex.getMessage())) >>
+            Ok("Oops, something went wrong, please try again later.")
+end EmailService
+```
+
+### 3.1.4 Creating our FinalService
+
+Let's create a FinalService.scala file and add the following contents
+
+```scala
+package com.rockthejvm
+
+class FinalService(generator: String, id: Int):
```

Comment:
Let's call this something different, like OtpInteractiveService or something

Reply:
Updated